pkg/policy/mapstate.go

```diff
@@ -858,28 +762,53 @@ func (ms *mapState) insertWithChanges(key Key, entry mapStateEntry, features pol
 // deleted.
 //
 // Incremental changes performed are recorded in 'changes'.
-func (ms *mapState) denyPreferredInsertWithChanges(newKey Key, newEntry mapStateEntry, features policyFeatures, changes ChangeState) {
-	// Bail if covered by a deny key
-	if features.contains(denyRules) {
+func (ms *mapState) insertWithChanges(newKey Key, newEntry mapStateEntry, features policyFeatures, changes ChangeState) {
+	if newEntry.IsDeny() {
+		// Bail if covered by a key of higher precedence, or by a deny key of the same
+		// precedence
 		for k, v := range ms.BroaderOrEqualKeys(newKey) {
-			// Identical deny key needs to be added to merge their entries.
-			if v.IsDeny && !(newEntry.IsDeny && k == newKey) {
+			if v.Precedence > newEntry.Precedence ||
+				// New deny entry is also bailed due to different covering deny key
+				// of the same precedence, equal keys need to be merged
+				v.Precedence == newEntry.Precedence && k != newKey {
```

Comment:
Does this need grouping to retain the current behavior?

```suggestion
				(v.Precedence == newEntry.Precedence && k != newKey) {
```

Reply:
If yes, there are similar examples lower in the function.

Reply:
I think these are now moot after I simplified this.